mypy/checker.py

```diff
@@ -1715,15 +1797,32 @@ def check_incompatible_property_override(self, e: Decorator) -> None:
 
     def visit_with_stmt(self, s: WithStmt) -> Type:
         echk = self.expr_checker
+        if s.is_async:
+            m_enter = '__aenter__'
+            m_exit = '__aexit__'
+        else:
+            m_enter = '__enter__'
+            m_exit = '__exit__'
         for expr, target in zip(s.expr, s.target):
             ctx = self.accept(expr)
-            enter = echk.analyze_external_member_access('__enter__', ctx, expr)
+            enter = echk.analyze_external_member_access(m_enter, ctx, expr)
             obj = echk.check_call(enter, [], [], expr)[0]
+            if s.is_async:
+                self.check_subtype(obj, self.named_type('typing.Awaitable'), expr)
             if target:
+                if s.is_async:
+                    # XXX TODO What if it's a subclass of Awaitable?
+                    if (isinstance(obj, Instance) and len(obj.args) == 1
+                            and obj.type.fullname() == 'typing.Awaitable'):
+                        obj = obj.args[0]
+                    else:
+                        obj = AnyType()
```

Comment:
Here's a proposal. typing.Awaitable already defines `__await__` as a generator, in `typing.pyi` (the implementation is structured differently):
```python
class Awaitable(Generic[_T_co]):
    @abstractmethod
    def __await__(self) -> Generator[AbstractFuture[_T_co], Any, _T_co]:...
```
So we can assume that anything that subclasses Awaitable must still implement `__await__` matching this same signature, so we can call get_generator_return_type() on it to extract the substibution for `_T_co`.

(FWIW I think the AbstractFuture for the yielded value is overbearing and would like to replace it with Any. But not in this diff, and it doesn't affect the thing here.)